README.md
# bevy_text_mesh &emsp; [![Build Status]][actions] [![Latest Version]][crates.io] [![Docs Version]][docs]

[build status]: https://img.shields.io/github/actions/workflow/status/blaind/bevy_text_mesh/ci.yml?branch=main
[actions]: https://github.com/blaind/bevy_text_mesh/actions?query=branch%3Amain
[latest version]: https://img.shields.io/crates/v/bevy_text_mesh.svg
[crates.io]: https://crates.io/crates/bevy_text_mesh
[docs version]: https://docs.rs/bevy_text_mesh/badge.svg
[docs]: https://docs.rs/bevy_text_mesh

A bevy 3D text mesh generator plugin for displaying text in 3D scenes

![Example](docs/highlight.webp)

The text mesh is generated at runtime from runtime-tessellated (and cached) TrueType font glyphs. Tessellation of glyphs is done with C-based [github.com/fetisov/ttf2mesh](https://github.com/fetisov/ttf2mesh/) library that is being interfaced through Rust-based FFI API (see [ttf2glyph-rs](https://crates.io/crates/ttf2mesh)).

## Known limitations

Consider this as a preview of the plugin for gathering feedback about the API:

- **The API will change in future - still iterating**
- Multiple `TextMesh` configuration fields are not implemented yet, see example below
- Text color update is not implemented yet
- Spacing of characters are incorrect
- Mesh cache purging is not implemented - this implementation will leak memory (see [#2](https://github.com/blaind/bevy_text_mesh/issues/2))
- WASM builds are not supported (see [#11](https://github.com/blaind/bevy_text_mesh/issues/11))

## Bevy versions support table

| bevy | bevy_text_mesh |
| ---- | -------------- |
| 0.12 | 0.9.0          |
| 0.11 | 0.7.0          |
| 0.10 | 0.6.0          |
| 0.9  | 0.5.0          |
| 0.8  | 0.4.0          |
| 0.7  | 0.2.0          |
| 0.6  | 0.1.0          |
| 0.5  | 0.0.2          |

## Usage

## Prequisites

Prequisites (for compiling [ttf2mesh-rs](https://crates.io/crates/ttf2mesh)):

    apt-get install build-essential patch

## Running the examples

See the [examples](/examples) -folder.

```
git clone https://github.com/blaind/bevy_text_mesh.git
cd bevy_text_mesh
cargo run --example 3d_scene --release # or
cargo run --example performance --release
```

## Integrating to your Bevy App

Add to Cargo.toml:

```
[dependencies]
bevy_text_mesh = "0.9.0"
```

Include the library:

```rust
use bevy_text_mesh::prelude::*;
```

Second, add a `TextMeshPlugin` to your app:

```rust
App::new()
    ...
    .add_plugins(TextMeshPlugin)
    ...;
```

Then, add the desired TrueType-fonts (with suffix `.ttf`) into your assets folder, a good convention is to store them to `assets/fonts` folder.

For example, see Fira fonts. Please read also their [LICENSE](https://github.com/mozilla/Fira/blob/master/LICENSE).

    mkdir -p assets/fonts
    wget https://github.com/mozilla/Fira/raw/master/ttf/FiraSans-Medium.ttf -O assets/fonts/FiraSans-Medium.ttf

Next, you are ready to spawn a text in your scene at a system:

First, load a font asset:

```rust
let font: Handle<TextMeshFont> = asset_server.load("fonts/FiraSans-Medium.ttf#mesh");
```

Then, spawn a textmesh bundle:

```rust
commands.spawn(TextMeshBundle {
    text_mesh: TextMesh::new_with_color("Hello Bevy", font, Color::rgb(1., 1., 0.)),
    transform: Transform::from_xyz(-1., 1.75, 0.),
    ..Default::default()
});
```

Or with expanded syntax:

```rust
commands.spawn(TextMeshBundle {
    text_mesh: TextMesh {
        text: String::from("Hello Bevy!"),
        style: TextMeshStyle {
            font,
            font_size: SizeUnit::NonStandard(36.),
            color: Color::rgb(1.0, 1.0, 0.0),
            font_style: FontStyle::UPPERCASE, // only UPPERCASE & LOWERCASE implemented currently
            mesh_quality: Quality::Low,
            ..Default::default()
        },
        alignment: TextMeshAlignment {
            vertical: VerticalAlign::Top, // FUNCTIONALITY NOT IMPLEMENTED YET - NO EFFECT
            horizontal: HorizontalAlign::Left, // FUNCTIONALITY NOT IMPLEMENTED YET - NO EFFECT
            ..Default::default()
        },
        size: TextMeshSize {
            width: SizeUnit::NonStandard(135.),       // partially implemented
            height: SizeUnit::NonStandard(50.),       // partially implemented
            depth: Some(SizeUnit::NonStandard(50.0)), // must be > 0 currently, 2d mesh not supported yet
            wrapping: true,                           // partially implemented
            overflow: false,                          // NOT IMPLEMENTED YET
            ..Default::default()
        },
        ..Default::default()
    },
    transform: Transform {
        translation: Vec3::new(-1., 1.75, 0.),
        ..Default::default()
    },
    ..Default::default()
});
```

## License

Licensed under <a href="LICENSE">MIT license</a>

### Contribution

Unless you explicitly state otherwise, any contribution intentionally submitted
for inclusion in the software by you, shall be licensed as above, without any additional terms or conditions.


CHANGELOG.md
# Changelog

[git_tag_comparison]: https://github.com/blaind/bevy_text_mesh/compare/v0.9.0...main

## Version 0.9.0 (2023-11-21)

[Compare changelog](https://github.com/blaind/bevy_text_mesh/compare/v0.8.0...v0.9.0)

### Changed

- [Breaking: reverted the `#mesh` suffix for font loading - the suffix is required again][34]

## Version 0.8.0 (2023-11-19)

[Compare changelog](https://github.com/blaind/bevy_text_mesh/compare/v0.7.0...v0.8.0)

### Changed

- [Upgrade bevy to 0.12.0][32]
- Breaking: font must no longer be loaded with `#mesh` suffix

## Version 0.7.0 (2023-08-05)

[Compare changelog](https://github.com/blaind/bevy_text_mesh/compare/v0.6.0...v0.7.0)

This version was never published to crates.io.

### Changed

- [Upgrade bevy to 0.11.0][29]

## Version 0.6.0 (2023-04-11)

[Compare changelog](https://github.com/blaind/bevy_text_mesh/compare/v0.5.0...v0.6.0)

### Changed

- Upgrade bevy to 0.10.0
- Upgrade bitflags to 2.1.0

## Version 0.5.0 (2022-11-13)

[Compare changelog](https://github.com/blaind/bevy_text_mesh/compare/v0.4.0...v0.5.0)

### Changed

- Upgrade bevy to 0.9.0

## Version 0.4.0 (2022-10-24)

[Compare changelog](https://github.com/blaind/bevy_text_mesh/compare/v0.3.0...v0.4.0)

### Changed

- [Breaking: use #mesh label for font loading to allow interop with bevy `.ttf` AssetLoader][15]

## Version 0.3.0 (2022-08-25)

[Compare changelog](https://github.com/blaind/bevy_text_mesh/compare/v0.2.0...v0.3.0)

### Changed

- Upgrade bevy to 0.8.0

## Version 0.2.0 (2022-04-17)

[Compare changelog](https://github.com/blaind/bevy_text_mesh/compare/v0.1.0...v0.2.0)

### Changed

- Upgrade bevy to 0.7.0

[15]: https://github.com/blaind/bevy_text_mesh/pull/15
[29]: https://github.com/blaind/bevy_text_mesh/pull/29
[32]: https://github.com/blaind/bevy_text_mesh/pull/32
[34]: https://github.com/blaind/bevy_text_mesh/pull/34


examples/3d_scene.rs
use std::time::Duration;

use bevy::{prelude::*, render::camera::Camera};
use bevy_text_mesh::prelude::*;

fn main() {
    App::new()
        .insert_resource(Msaa::Sample4)
        .add_plugins((DefaultPlugins, TextMeshPlugin))
        .add_systems(Startup, (setup, setup_text_mesh.after(setup)))
        .add_systems(Update, (update_text_mesh, rotate_camera))
        .run();
}

fn setup_text_mesh(mut commands: Commands, asset_server: Res<AssetServer>) {
    let font: Handle<TextMeshFont> = asset_server.load("fonts/FiraMono-Medium.ttf#mesh");

    commands.spawn(TextMeshBundle {
        text_mesh: TextMesh {
            text: String::from("Time since startup"),
            style: TextMeshStyle {
                font: font.clone(),
                font_size: SizeUnit::NonStandard(9.),
                color: Color::rgb(0.0, 0.0, 0.0),
                ..Default::default()
            },
            size: TextMeshSize {
                ..Default::default()
            },
            ..Default::default()
        },
        transform: Transform {
            translation: Vec3::new(-1., 1.75, 0.),
            ..Default::default()
        },
        ..Default::default()
    });

    commands
        .spawn(TextMeshBundle {
            text_mesh: TextMesh {
                text: String::from("0"),
                style: TextMeshStyle {
                    font: font.clone(),
                    font_size: SizeUnit::NonStandard(36.),
                    color: Color::rgb(0.0, 1.0, 0.0),
                    mesh_quality: Quality::Custom(128),
                    ..Default::default()
                },
                ..Default::default()
            },
            transform: Transform {
                translation: Vec3::new(-1., 1.3, 0.),
                ..Default::default()
            },
            ..Default::default()
        })
        .insert(EngineTime);

    commands.insert_resource(UpdateTimer {
        timer: Timer::new(Duration::from_millis(100), TimerMode::Repeating),
    });
}

#[derive(Resource)]
struct UpdateTimer {
    timer: Timer,
}

#[derive(Component)]
struct EngineTime;

fn update_text_mesh(
    time: Res<Time>,
    mut text_meshes: Query<&mut TextMesh, With<EngineTime>>,
    mut timer: ResMut<UpdateTimer>,
) {
    if timer.timer.tick(time.delta()).just_finished() {
        for mut text_mesh in text_meshes.iter_mut() {
            let updated_text = String::from(format!("Time = {:.3}", time.elapsed_seconds_f64()));

            if text_mesh.text != updated_text {
                text_mesh.text = updated_text;
            }
        }
    }
}

fn rotate_camera(mut camera: Query<&mut Transform, With<Camera>>, time: Res<Time>) {
    for mut camera in camera.iter_mut() {
        let angle = time.elapsed_seconds_f64() as f32 / 2. + 1.55 * std::f32::consts::PI;

        let distance = 3.5;

        camera.translation = Vec3::new(
            angle.sin() as f32 * distance,
            camera.translation.y,
            angle.cos() as f32 * distance,
        );

        *camera = camera.looking_at(Vec3::new(0.0, 1.5, 0.), Vec3::Y);
    }
}

/// set up a simple 3D scene
fn setup(
    mut commands: Commands,
    mut meshes: ResMut<Assets<Mesh>>,
    mut materials: ResMut<Assets<StandardMaterial>>,
) {
    commands.spawn(PbrBundle {
        mesh: meshes.add(Mesh::from(shape::Plane::from_size(5.0))),
        material: materials.add(Color::rgb(0.3, 0.5, 0.3).into()),
        ..Default::default()
    });
    commands.spawn(PbrBundle {
        mesh: meshes.add(Mesh::from(shape::Cube { size: 1.0 })),
        material: materials.add(Color::rgb(0.8, 0.7, 0.6).into()),
        transform: Transform::from_xyz(0.0, 0.5, 0.0),
        ..Default::default()
    });
    commands.spawn(PointLightBundle {
        transform: Transform::from_xyz(4.0, 8.0, 4.0),
        ..Default::default()
    });
    commands.spawn(Camera3dBundle {
        transform: Transform::from_xyz(-2.0, 2.5, 5.0).looking_at(Vec3::ZERO, Vec3::Y),
        ..Default::default()
    });
}
